packages/react-query/src/utils/structuralSharing.ts

```diff
@@ -0,0 +1,132 @@
+ */
+export function createStructuralSharingFunction(
+  equalityFn: (a: any, b: any) => boolean,
+) {
+  const structuralSharingFunction = (a: any, b: any): any => {
+    if (equalityFn(a, b)) {
+      return a;
+    }
+
+    const array = isPlainArray(a) && isPlainArray(b);
+    const plainObject = isPlainObject(a) && isPlainObject(b);
+    const map = a instanceof Map && b instanceof Map;
+    const set = a instanceof Set && b instanceof Set;
+
+    if (array || plainObject || map) {
+      const aItems = array
+        ? a
+        : plainObject
+          ? Object.keys(a)
+          : Array.from(a.keys());
+      const aSize = aItems.length;
+      const bItems = array
+        ? b
+        : plainObject
+          ? Object.keys(b)
+          : Array.from(b.keys());
+      const bSize = bItems.length;
+      const copy: any = array ? [] : plainObject ? {} : new Map();
+
+      let equalItems = 0;
+
+      for (let i = 0; i < bSize; i++) {
+        const key = array ? i : bItems[i];
+        if (
+          !array &&
+          a[key] === undefined &&
+          b[key] === undefined &&
+          aItems.includes(key)
+        ) {
+          copy[key] = undefined;
+          equalItems++;
+        } else {
+          copy[key] = structuralSharingFunction(a[key], b[key]);
+          if (copy[key] === a[key] && a[key] !== undefined) {
+            equalItems++;
+          }
+        }
+      }
+
+      return aSize === bSize && equalItems === aSize ? a : copy;
+    }
+
+    if (set) {
+      const copy: any = new Set();
+      let equalItems = 0;
+      if (a.size !== b.size) return false;
+      const aItems = Array.from(a.values());
+      for (const bItem of b.values()) {
+        // We're doing a shallow comparison here, not a deep one.
+        const aItem = aItems.find((aItem) => equalityFn(aItem, bItem));
+        if (!aItem) {
+          copy.add(bItem);
+        } else {
+          copy.add(aItem);
+          equalItems++;
+        }
+      }
+      return a.size === b.size && equalItems === a.size ? a : copy;
+    }
+
+    return b;
+  };
+
+  return structuralSharingFunction;
+}
+
+export const defaultStructuralSharingFunction =
+  createStructuralSharingFunction(isEqual);
+
+function isEqual(a: any, b: any) {
+  if (a === b) return true;
+  if (Object.is(a, b)) return true;
+  if (a.constructor !== b.constructor) return false;
+  if (a.constructor === RegExp)
+    return a.source === b.source && a.flags === b.flags;
+  if (a.valueOf !== Object.prototype.valueOf)
+    return a.valueOf() === b.valueOf();
+  if (a.toString !== Object.prototype.toString)
+    return a.toString() === b.toString();
+  return false;
+}
+
+function isPlainArray(value: unknown) {
+  return Array.isArray(value) && value.length === Object.keys(value).length;
+}
+
+// Copied from: https://github.com/jonschlinkert/is-plain-object
+function isPlainObject(o: any): o is object {
+  if (!hasObjectPrototype(o)) {
+    return false;
+  }
```

Comment:
We have a much simplified version here:

https://github.com/trpc/trpc/blob/3957a7dc564b14f52ce6ba5078490e73b6b6be98/packages/server/src/unstable-core-do-not-import/utils.ts#L26-L32

Which one is wrong? 😅

Reply:
This comes for RQ and I believe most of it aims to avoid edge cases like this one https://github.com/TanStack/query/pull/7357.
Because we do `copy = {}` if there's a custom constructor we lose something.